bson/src/main/org/bson/codecs/pojo/ClassModel.java

```diff
@@ -44,13 +45,13 @@
         this.name = clazz.getSimpleName();
         this.type = clazz;
         this.hasTypeParameters = clazz.getTypeParameters().length > 0;
-        this.propertyNameToTypeParameterMap = propertyNameToTypeParameterMap;
+        this.propertyNameToTypeParameterMap = Collections.unmodifiableMap(propertyNameToTypeParameterMap);
```

Comment:
It can't be that some piece of code is accessing this class's instance variables and modifying them.  If that were the case, then we'd now be seeing `UnsupportedOperationException` from that code, now that the instance variable are wrapped like this.

Given that,  then it must be that some piece of calling code is retaining a reference to the parameters and then modifying them.  In which case what we really need to do is make a defensive copy of the parameters. `Collections.unmodifiableMap/List` doesn't do that for you.